### PR TITLE
Add USB Prorotyping Mode

### PR DESCRIPTION
**Pull request by cnlohr**

### CODE REVIEW

File: main/swadgeMode.h [hunk 1]
```diff
@@ -126,4 +126,7 @@ typedef struct _swadgeMode
     void (*fnEspNowSendCb)(const uint8_t* mac_addr, esp_now_send_status_t status);
 } swadgeMode;
 
-#endif
\ No newline at end of file
+void overrideToSwadgeMode( swadgeMode* mode );
+void switchToSwadgeMode(uint8_t mode);
```
Comment (AEFeinstein) [inline main/swadgeMode.h:130]: Fine here, but remove it from `mode_main_menu.h` please. No one needs duplicates!

File: main/swadge_esp32.c [hunk 2]
```diff
@@ -146,15 +154,20 @@ uint16_t tud_hid_get_report_cb(uint8_t itf __attribute__((unused)),
...
+    hid_report_type_t report_type,
+    uint8_t const* buffer,
+    uint16_t bufsize )
 {
-    ;
+    if( report_id == 170 )
+        handle_advanced_usb_control_set( bufsize, buffer );
 }
 
+#include "soc/dport_access.h"
```
Comment (AEFeinstein) [inline main/swadge_esp32.c:167]: Move to be with other `#includes` please.

File: tools/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
+Be sure to put the following in: `/etc/udev/rules.d/99-magfest.rules` and execute `sudo udevadm control --reload-rules && sudo udevadm trigger`
```
Comment (AEFeinstein) [inline tools/README.md:2]: Checking if/how this works on windows would be nice too.

File: tools/flash_test/flash_test.c [hunk 4]
```diff
@@ -0,0 +1,201 @@
+#include <stdio.h>
+#include <stdint.h>
+
+#include "../hidapi.h"
+#include "../hidapi.c"
+
+#define VID 0x303a
```
Comment (AEFeinstein) [inline tools/flash_test/flash_test.c:7]: Note to self, this is Espressif's VID

File: tools/flash_test/flash_test.c [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		
+		printf( "Writing file %s to 0x%x (length: %d)\n", argv[2], flash_start, bloblen );
+		int blobkclen = ( bloblen + alignlen - 1 ) / alignlen * alignlen;
+		uint8_t * blob;
+		{
+			FILE * binary_blob = fopen( argv[2], "r" );
+			if( !binary_blob )
+			{
+				fprintf( stderr, "ERROR: Could not open %s.\n", argv[2] );
+				return -96;
```
Comment (AEFeinstein) [inline tools/flash_test/flash_test.c:69]: Where are -96 and -97 coming from?

**Comment by AEFeinstein** [comment]
Emulator compilation is broken, and I'm not going to merge until that's fixed. I think you'll probably preprocess out some USB stuff rather than stubbing it.

In general I'd like to see more comments (doxygen and inline), and enums (rather than magic numbers) just to have a better idea of what's going on. I also reserve the right to `astyle` or `clang-format` this at some point.

File: components/hdw-tft/hdw-tft.c [hunk 6]
```diff
@@ -377,7 +387,7 @@ void initTFT(display_t * disp, spi_host_device_t spiHost, gpio_num_t sclk,
 
 #if defined(CONFIG_ST7735_160x80)
     ESP_ERROR_CHECK(esp_lcd_new_panel_st7735(io_handle, &panel_config, &panel_handle));
-#elif defined(CONFIG_ST7789_240x135) || defined(CONFIG_ST7789_240x240)
+#elif defined(CONFIG_ST7789_240x135) || defined(CONFIG_ST7789_240x240) || defined(CONFIG_ST7735_128x160)
```
Comment (ADC-AFeinstein) [inline components/hdw-tft/hdw-tft.c:390]: Did you mean to put `CONFIG_ST7735_128x160` where it calls `esp_lcd_panel_st7789()` and not above where it calls `esp_lcd_panel_st7735()`?

Comment (AEFeinstein) [inline components/hdw-tft/hdw-tft.c:390]: Did you mean to put `CONFIG_ST7735_128x160` where it calls `esp_lcd_panel_st7789()` and not above where it calls `esp_lcd_panel_st7735()`?

**Comment by cnlohr** [comment]
> What are the odds this works on windows too?

This will work on Windows, but I don't plan to add support for it until weekend after this coming one.

File: tools/README.md (hunk 3 above)
Comment (cnlohr) [reply tools/README.md:2]: Will work on Windows, just not right now.

File: tools/sandbox_test/Makefile [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+SYSELF:=../../build/swadge-esp32.elf
+CFLAGS:=-g -O0
+
+build : 
+	mkdir -p build
+
+sandbox.o : buildhelp sandbox.c $(SYSELF) build
+	./buildhelp $(SYSELF)
+	xtensa-esp32s2-elf-objdump -s build/sandbox.o > build/debug_sandbox_s.txt
+	xtensa-esp32s2-elf-objdump -s build/sandbox.o > build/debug_sandbox_t.txt
```
[the remark this answers is not in the document]
Comment (cnlohr) [reply tools/sandbox_test/Makefile:12]: I did! Will fix tonight.

**Comment by cnlohr** [comment]
> Emulator compilation is broken, and I'm not going to merge until that's fixed. I think you'll probably preprocess out some USB stuff rather than stubbing it.
> 
> In general I'd like to see more comments (doxygen and inline), and enums (rather than magic numbers) just to have a better idea of what's going on. I also reserve the right to `astyle` or `clang-format` this at some point.

Can do.

File: components/hdw-tft/hdw-tft.c (hunk 6 above)
Comment (cnlohr) [reply components/hdw-tft/hdw-tft.c:390]: I will review this tonight.

File: tools/flash_test/flash_test.c (hunk 5 above)
Comment (AEFeinstein) [reply tools/flash_test/flash_test.c:69]: Ok. I always prefer something marginally less arbitrary, like an enum, but it is what it is.

File: components/hdw-tft/hdw-tft.c (hunk 6 above)
Comment (cnlohr) [reply components/hdw-tft/hdw-tft.c:390]: I believe it worked both ways.  I don't recall why I settled on this.  It's not actually a 7735

File: components/hdw-spiffs/spiffs_json.c [hunk 8]
```diff
@@ -23,7 +22,7 @@ char * loadJson(char * name)
     size_t sz;
     if(!spiffsReadFile(name, &buf, &sz))
     {
-        ESP_LOGE("JSON", "Failed to read %s", name);
+        printf("Failed to read %s", name);
```
Comment (AEFeinstein) [inline components/hdw-spiffs/spiffs_json.c:25]: Any particular reason you replaced `ESP_LOGE()` with `printf()`? My intent was to turn these all off in final firmware via `sdkconfig`

File: components/hdw-tft/hdw-tft.c (hunk 6 above)
Comment (AEFeinstein) [reply components/hdw-tft/hdw-tft.c:390]: Then the preprocessor check probably shouldn't say `7735`...

**Comment by AEFeinstein** [comment]
Emulator compilation is still broken:
```
main/advanced_usb_control.c:5:10: fatal error: esp_partition.h: No such file or directory
    5 | #include "esp_partition.h"
      |          ^~~~~~~~~~~~~~~~~
```
Would it be easier for me to merge, then fix it myself?

File: components/hdw-spiffs/spiffs_json.c (hunk 8 above)
Comment (cnlohr) [reply components/hdw-spiffs/spiffs_json.c:25]: EMU/Non-EMU operation.  How else would you want to handle it?

**Comment by cnlohr** [comment]
> Emulator compilation is still broken:
> 
> ```
> main/advanced_usb_control.c:5:10: fatal error: esp_partition.h: No such file or directory
>     5 | #include "esp_partition.h"
>       |          ^~~~~~~~~~~~~~~~~
> ```
> 
> Would it be easier for me to merge, then fix it myself?

I am not getting this error. Where do you get it?  I'm not including main/advanced_usb_control.c in my build tree for the emu?

File: components/hdw-spiffs/spiffs_json.c (hunk 8 above)
Comment (AEFeinstein) [reply components/hdw-spiffs/spiffs_json.c:25]: The emu should already route those prints to printf anyway...

**Pull request merged**